core/src/connection_reuse.rs

```diff
@@ -40,193 +40,543 @@
+    }
+}
+
+enum PeerState<M>
+where
+    M: StreamMuxer + Clone,
+{
+    /// Connection is active and can be used to open substreams.
+    Active {
+        /// The muxer to open new substreams.
+        muxer: M,
+        /// Id of the listener that created this connection, or `None` if it was opened by a
+        /// dialer.
+        listener_id: Option<usize>,
+        /// whether this connection was closed
+        closed: Arc<AtomicBool>
+    },
+
+    /// Connection is pending.
+    // TODO: stronger Future type
+    Pending {
+        /// Future that produces the muxer.
+        future: Box<Future<Item = (M, Multiaddr), Error = IoError>>,
+        /// All the tasks to notify when `future` resolves.
+        notify: HashMap<usize, task::Task, UsizeHasherBuilder>,
+    },
+
+    /// An earlier connection attempt errored.
+    Errored(IoError),
+}
+
+impl<M> PeerState<M>
+where
+    M: StreamMuxer + Clone
+{
+    /// resets the connections if the given PeerState is active and has
+    /// substreams. Consume the PeerState while doing so.
+    fn reset_conn(self) {
+        if let PeerState::Active { closed, .. } = self {
+            closed.store(false, Ordering::Release);
+        }
+    }
+}
+
+/// Struct shared between most of the `ConnectionReuse` infrastructure.
+/// Knows about all connections and their current state, allows one to poll for
+/// incoming and outbound connections, while managing all state-transitions that
+/// might occur automatically.
+// #[derive(Clone)]
+struct ConnectionsManager<T, C>
+where
+    T: Transport,
     C: ConnectionUpgrade<T::Output, T::MultiaddrFuture>,
-    C::Output: StreamMuxer,
+    C::Output: StreamMuxer + Clone,
 {
-    // Underlying transport and connection upgrade for when we need to dial or listen.
-    inner: UpgradedNode<T, C>,
+    /// Underlying transport and connection upgrade, used when we need to dial or listen.
+    transport: UpgradedNode<T, C>,
+
+    /// All the connections that were opened, whether successful and/or active or not.
+    // TODO: this will grow forever
+    connections: Mutex<FnvHashMap<Multiaddr, PeerState<C::Output>>>,
+
+    /// Tasks to notify when one or more new elements were added to `connections`.
+    notify_on_new_connec: Mutex<HashMap<usize, task::Task, UsizeHasherBuilder>>,
 
-    // Struct shared between most of the `ConnectionReuse` infrastructure.
-    shared: Arc<Mutex<Shared<C::Output>>>,
+    /// Counter giving us the next listener_id
+    listener_counter: Counter,
 }
 
-struct Shared<M>
+impl<T, C> ConnectionsManager<T, C>
 where
-    M: StreamMuxer,
+    T: Transport,
+    C: ConnectionUpgrade<T::Output, T::MultiaddrFuture>,
+    C::Output: StreamMuxer + Clone,
 {
-    // List of active muxers.
-    active_connections: FnvHashMap<Multiaddr, M>,
+    /// Notify all `Tasks` in `self.notify_on_new_connec` that a new connection was established
+    /// consume the tasks in that process
+    fn new_con_notify(&self) {
```

Comment:
I commented about this in private, but I'm really not a fan of locking a mutex in a function that is intended to be called from another function which can potentially lock a mutex as well.
This greatly increases the difficulty of the reviewer if you want to make sure that there is no deadlock.

Reply:
> I commented about this in private, but I'm really not a fan of locking a mutex in a function that is intended to be called from another function which can potentially lock a mutex as well.

While I agree on the general merit of this statement, I don't think this applies here: There are two distinct `Mutex`s, each one is clearly bound to corresponding function calls. I don't really see how it gives us any benefit if instead, we'd `Mutex` the entire `ConnectionManager` (what we had before) and therefore prevent anyone using the `new_notify_registration` (the only function actually used from the outside) until we are done with the respective `poll_`s. Even in that scenario, if `task.1.notify` _would_ cause an immediate `poll` (which is the only side-effect function and we know _it doesn't_ as this wake-up is only queued) we'd deadlock. 

I totally agree with you that `Mutex`s make the reviewer harder, as you need to make sure there isn't any side-effect code within the `lock`ed code that might cause a `deadlock`. Which is exactly why I try to keep the code between locks as constraint, short and side-effect-free as possible.

I am fine with moving this lock+drain on the respective position in the code base (as it is only used once) but I see little to no merit in putting the entire object into a mutex just for these two clearly independent locks... If you insist however, I can also do that.